src/common/ScreenExporter.cpp

```
@@ -0,0 +1,198 @@
+    int rowbytes = width * 3;
+
+    std::vector<uInt8> buffer((rowbytes + 1) * height, 0);
+    uInt8* buf_ptr = &buffer[0];
+
+    for(int i = 0; i < height; i++) {
+        *buf_ptr++ = 0;                  // first byte of row is filter type
+        for(int j = 0; j < dataWidth; j++) {
+            int r, g, b;
+
+            palette.getRGB(screen.getArray()[i * dataWidth + j], r, g, b);
+            // Double the pixel width, if so desired
+            int jj = doubleWidth ? 2 * j : j;
+
+            buf_ptr[jj * 3 + 0] = r;
+            buf_ptr[jj * 3 + 1] = g;
+            buf_ptr[jj * 3 + 2] = b;
+            
+            if (doubleWidth) {
+                
+                jj = jj + 1;
+
+                buf_ptr[jj * 3 + 0] = r;
+                buf_ptr[jj * 3 + 1] = g;
+                buf_ptr[jj * 3 + 2] = b;
+            }
+        }
+        buf_ptr += rowbytes;                 // add pitch
+    }
+
+    // Compress the data with zlib
+    uLongf compmemsize = (uLongf)((height * (width + 1) * 3 + 1) + 12);
+    std::vector<uInt8> compmem(compmemsize, 0);
+    
+    if((compress(&compmem[0], &compmemsize, &buffer[0], height * (width * 3 + 1)) != Z_OK)) {
+
+        // @todo -- throw a proper exception
+        std::cerr << "Error: Couldn't compress PNG" << std::endl;
+        return;
+    }
+
+    // Write the compressed framebuffer data
+    writePNGChunk(out, "IDAT", &compmem[0], compmemsize);
+}
+
+
+static void writePNGEnd(std::ofstream &out) {
+
+    // Finish up
+    writePNGChunk(out, "IEND", 0, 0);
+}
+
+ScreenExporter::ScreenExporter(ColourPalette &palette):
+    m_palette(palette),
+    m_frame_number(0),
+    m_frame_field_width(6),
+    m_basename("") {
+}
+
+
+ScreenExporter::ScreenExporter(ColourPalette &palette, const std::string &basename):
+    m_palette(palette),
+    m_frame_number(0),
+    m_frame_field_width(6),
+    m_basename(basename) {
+
+    // MGB: Should we strip '.png' from filename if present?
```

Maybe an assert could work here? If a user provides a bad, path it's not going to be able to save anyway, so better just to let them know + fail early?